### DISCUSSION TITLE
Audio Mix/DSP Engine

### DISCUSSION THREAD

**Discussion by stevefan1999-personal** [ppy/osu#24456] [Ideas]
I'm trying to make a game mode with 6 buttons and 2 knobs (should be obvious to anyone who spent some years in arcade rhythm games, you know exactly what I mean) that requires the use of special sound effects like Bitcrusher, Gate, Wobble and much more.  This means we will need to have a (soft) real-time mixing engine. I mean it does not need to have the level of precision typically seen in a DAW but still real time enough the interacting player won't notice any glitches or delays. 

Right now, all I saw was just a simple BQF low and high pass filter for beatmap selection and failing sound effect, but that's simply not enough.

Somewhat related: https://github.com/ppy/osu/pull/16771, but for mods only rather than game mode.

I will be planning on using NAudio for mixing: https://github.com/naudio/NAudio

**Comment by peppy** [comment]
If you're not using bass effects, you're kind of on your own. I'd highly recommend looking into BassFx as it will likely be much easier to implement than passing audio through NAudio for processing.
